doc/warm-reboot/NSF_Manager_HLD.md

```diff
@@ -0,0 +1,280 @@
+*   Phase 3: Trigger checkpointing
+*   Phase 4: Prepare and perform reboot
+
+
+##### Phase 1: Freeze Components & Wait for Switch Quiescence
+
+
+
+![alt_text](img/freeze.png)
+
+NSF Manager will send freeze notification to all registered components and will wait for the quiescence of only those components that have set _freeze = true_. Upon receiving the freeze notification, the component will complete processing the current request queue and stop generating new intents. Stopping new intents from being generated means that boundary components should stop processing requests from external components (external events) and all components should stop their periodic timers that generate new requests (internal events). For example:
+
+
+
+*   UMF will stop listening to new gRPC requests from the controller.
+*   P4RT will stop listening to new gRPC requests from the controller and stop packet I/O.
+*   xcvrd will stop listening to transceiver updates such as module presence.
+*   Syncd will stop listening to link state events from the BCM chip.
+*   Orchagent will stop periodic internal timers.
+*   BGP will stop exchanging packets with the peers.
+*   Teamd will stop exchanging LACP PDUs with the peers.
+
+After all components have been freezed, the switch would eventually reach a state wherein each component stops generating new events and thus the switch becomes quiescent. This is because:
+
+
+
+*   Switch boundaries that generate new events have been stopped.
+*   All timers that generate new events have been stopped.
+*   All components have completed processing their pending requests and thus there are no in-flight messages.
+
+After receiving the freeze notification, the components will update their quiescent state in STATE DB when they receive a new request (i.e. they are no longer quiescent) and when they complete processing their current request queue (i.e. they become quiescent). NSF Manager will monitor the quiescent state of all components in STATE DB to determine that the switch has become quiescent and thus further state changes won’t occur in the switch. If all components are in quiescent state then NSF Manager will declare that the switch has become quiescent and thus the switch has attained its final state. NSF Manager will wait for a period of time for the switch to become quiescent after which it will determine that warm reboot failed and abort the warm reboot operation.
+
+
+##### Phase 2: State Verification (Optional)
+
+
+
+![alt_text](img/state-verification.png)
+
+Since the switch is in quiescent state, this will be the final state of the switch before reboot. NSF Manager will trigger state verification to ensure that the switch is in a consistent state. Reconciling from an inconsistent state might cause traffic loss and thus it is important to ensure that the switch is in a consistent state before warm rebooting. NSF Manager will abort warm reboot operation if state verification fails.
```

Comment:
How is this phase different from what was done as last step in Phase 1? Also, please define `consistent state`.

Reply:
Phase 1 ensures that the switch reaches a stable state. Phase 2 (optional) audits the switch state. State verification enables auditing of the state of different components and ensuring that they are in an expected state. Consistency depends on the component. It can mean that the internal cache matches the switch state, or the switch intent matches the switch state etc. For example, BGP kernel routes match APPL DB information, oper-status in Linux host interface matches APPL DB information etc.